tokio-trace/tokio-trace-core/src/span.rs

```diff
@@ -1,27 +1,121 @@
-/// created, through the [`new_id`](::Subscriber::new_span_id) trait
+/// created, by the [`new_span`](::Subscriber::new_span) trait
 /// method. See the documentation for that method for more information on span
 /// ID generation.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub struct Span(u64);
+pub struct Id(u64);
+
+/// Attributes provided to a `Subscriber` describing a new span when it is
+/// created.
+#[derive(Debug)]
+pub struct Attributes<'a> {
+    metadata: &'a Metadata<'a>,
+    values: &'a field::ValueSet<'a>,
+    parent: Parent,
+}
+
+#[derive(Debug)]
+enum Parent {
+    /// The new span will be a root span.
+    Root,
+    /// The new span will be rooted in the current span.
+    Current,
+    /// The new span has an explicitly-specified parent.
+    Explicit(Id),
+}
 
-// ===== impl Id =====
+// ===== impl Span =====
 
-impl Span {
+impl Id {
     /// Constructs a new span ID from the given `u64`.
     pub fn from_u64(u: u64) -> Self {
-        Span(u)
+        Id(u)
     }
 
     /// Returns the span's ID as a  `u64`.
     pub fn into_u64(&self) -> u64 {
         self.0
     }
 }
+
+// ===== impl Attributes =====
+
+impl<'a> Attributes<'a> {
+    /// Returns `Attributes` describing a new child span of the current span,
+    /// with the provided metadata and values.
+    pub fn new(metadata: &'a Metadata<'a>, values: &'a field::ValueSet<'a>) -> Self {
+        Attributes {
+            metadata,
+            values,
+            parent: Parent::Current,
+        }
+    }
+
+    /// Returns `Attributes` describing a new span at the root of its own trace
+    /// tree, with the provided metadata and values.
+    pub fn new_root(metadata: &'a Metadata<'a>, values: &'a field::ValueSet<'a>) -> Self {
+        Attributes {
+            metadata,
+            values,
+            parent: Parent::Root,
+        }
+    }
+
+    /// Returns `Attributes` describing a new child span of the specified
+    /// parent span, with the provided metadata and values.
+    pub fn child_of(
+        parent: Id,
+        metadata: &'a Metadata<'a>,
+        values: &'a field::ValueSet<'a>,
+    ) -> Self {
+        Attributes {
+            metadata,
+            values,
+            parent: Parent::Explicit(parent),
+        }
+    }
+
+    /// Returns a reference to the new span's metadata.
+    pub fn metadata(&self) -> &Metadata<'a> {
+        self.metadata
+    }
+
+    /// Returns a reference to a `ValueSet` containing any values the new span
+    /// was created with.
+    pub fn values(&self) -> &field::ValueSet<'a> {
+        self.values
+    }
+
+    /// Returns true if the new span shoold be a root.
+    pub fn is_root(&self) -> bool {
+        match self.parent {
+            Parent::Root => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if the new span should be a child of the current span.
+    pub fn is_in_current(&self) -> bool {
```

Comment:
Maybe this should be `is_contextual` or something like that? If there is no "current span" it will become a root.

Reply:
Yeah, I had a hard time coming up with a name that actually communicated all of what this meant...I suppose it could be `has_explicit_parent` or something but that seemed a little wordy.

Reply:
Changed it to `is_contextual` in 5218cf628d885cf5d271760914953802745b6134 :)